src/pyproject_hooks/_impl.py

```diff
@@ -6,31 +6,51 @@
 from os.path import abspath
 from os.path import join as pjoin
 from subprocess import STDOUT, check_call, check_output
+from typing import TYPE_CHECKING, Any, Iterator, Mapping, Optional, Sequence
 
 from ._in_process import _in_proc_script_path
 
+if TYPE_CHECKING:
+    from typing import Protocol
 
-def write_json(obj, path, **kwargs):
+    class SubprocessRunner(Protocol):
+        """A protocol for the subprocess runner."""
+
+        def __call__(
+            self,
+            cmd: Sequence[str],
+            cwd: Optional[str] = None,
```

Comment:
```suggestion
            cmd: Sequence[StrPath],
            cwd: Optional[StrPath] = None,
```

This and the rest of the code that accepts file paths should accept PathLike objects. There's a couple ways to annotate this, but my preference is to add `from _typeshed import StrPath` under an `if TYPE_CHECKING` block and use that.

The underlying subprocess code technically supports bytes as well. You can use `StrOrBytes` path for the subprocess_runner parts if you prefer.

Reply:
[`_typing.StrPath`](https://github.com/python/typeshed/blob/3c872ca8fd875f2dc5fe5f5d771e35c58390cd0e/stdlib/_typeshed/__init__.pyi#L143) is considered [stable](https://github.com/python/typeshed/tree/main/stdlib/_typeshed#api-stability). It is equivalent to `Union[str, "os.PathLike[str]"]` if you prefer to define it manually.

Reply:
Let's do this in a follow up change.